File: src/main/java/dsl/NestingMappingRule.java
package dsl;

import dsl.util.PathMatcher;
import xyz.mahmoudahmed.adapter.DataNode;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

/**
 * Rule that nests flat fields into a structured object or array.
 */
public class NestingMappingRule implements MappingRule {
    private final String sourcePattern;
    private final String targetPath;
    private final String structureName;
    private final String fieldNameExtractor;
    private final BiFunction<String, Object, Object> valueTransformation;
    private final boolean isCollection;

    public NestingMappingRule(String sourcePattern, String targetPath, String structureName,
                              String fieldNameExtractor, BiFunction<String, Object, Object> valueTransformation,
                              boolean isCollection) {
        this.sourcePattern = sourcePattern;
        this.targetPath = targetPath;
        this.structureName = structureName;
        this.fieldNameExtractor = fieldNameExtractor;
        this.valueTransformation = valueTransformation;
        this.isCollection = isCollection;
    }

    @Override
    public void apply(MappingContext context, Object target) {
        if (!(target instanceof DataNode.ObjectNode)) {
            return;
        }

        DataNode sourceNode = (DataNode) context.getSourceData();
        DataNode.ObjectNode targetNode = (DataNode.ObjectNode) target;

        // Find all paths that match the pattern
        List<PathMatcher.PathMatch> matches = PathMatcher.findMatches(sourceNode, sourcePattern);

        // Create the nested structure
        if (isCollection) {
            createNestedCollection(matches, targetNode);
        } else {
            createNestedObject(matches, targetNode);
        }
    }

    private void createNestedCollection(List<PathMatcher.PathMatch> matches, DataNode.ObjectNode targetNode) {
        // Create a collection of items
        DataNode.ArrayNode arrayNode = new DataNode.ArrayNode();

        for (PathMatcher.PathMatch match : matches) {
            DataNode.ObjectNode itemNode = new DataNode.ObjectNode();

            // Extract the field name to use as a key or include in the item
            String fieldName = extractFieldName(match.getPath());

            // Transform the value if needed
            Object value = match.getValue();
            if (valueTransformation != null) {
                value = valueTransformation.apply(fieldName, value);
            }

            // Add the field name if specified
            if (fieldNameExtractor != null && !fieldNameExtractor.isEmpty()) {
                itemNode.getProperties().put(fieldNameExtractor, new DataNode.StringValue(fieldName));
            }

            // Add the value
            String valueField = fieldNameExtractor != null && !fieldNameExtractor.isEmpty() ?
                    "value" : fieldName;

            if (value instanceof DataNode) {
                itemNode.getProperties().put(valueField, (DataNode) value);
            } else {
                itemNode.getProperties().put(valueField, convertToDataNode(value));
            }

            // Add to collection
            arrayNode.getElements().add(itemNode);
        }

        // Set the array in the target
        setValueInPath(targetNode, targetPath + "." + structureName, arrayNode);
    }

    private void createNestedObject(List<PathMatcher.PathMatch> matches, DataNode.ObjectNode targetNode) {
        // Create a single object with all the fields
        DataNode.ObjectNode nestedObject = new DataNode.ObjectNode();

        for (PathMatcher.PathMatch match : matches) {
            // Extract the field name
            String fieldName = extractFieldName(match.getPath());

            // Transform the value if needed
            Object value = match.getValue();
            if (valueTransformation != null) {
                value = valueTransformation.apply(fieldName, value);
            }

            // Add to nested object
            if (value instanceof DataNode) {
                nestedObject.getProperties().put(fieldName, (DataNode) value);
            } else {
                nestedObject.getProperties().put(fieldName, convertToDataNode(value));
            }
        }

        // Set the object in the target
        setValueInPath(targetNode, targetPath + "." + structureName, nestedObject);
    }

    private String extractFieldName(String path) {
        // Extract the part of the path that matches the wildcard in the pattern
        if (sourcePattern.contains("*")) {
            String prefix = sourcePattern.substring(0, sourcePattern.indexOf("*"));
            String suffix = sourcePattern.substring(sourcePattern.indexOf("*") + 1);

            if (path.startsWith(prefix) &&
                    (suffix.isEmpty() || path.endsWith(suffix))) {

                int startIndex = prefix.length();
                int endIndex = suffix.isEmpty() ? path.length() : path.length() - suffix.length();
                return path.substring(startIndex, endIndex);
            }
        }

        // Default: just use the last segment of the path
        int lastDot = path.lastIndexOf('.');
        return lastDot >= 0 ? path.substring(lastDot + 1) : path;
    }

    private void setValueInPath(DataNode.ObjectNode targetNode, String path, Object value) {
        String[] parts = path.split("\\.");
        DataNode.ObjectNode current = targetNode;

        // Navigate to the parent node
        for (int i = 0; i < parts.length - 1; i++) {
            String part = parts[i];
            DataNode child = current.get(part);

            if (child == null || !(child instanceof DataNode.ObjectNode)) {
                DataNode.ObjectNode newNode = new DataNode.ObjectNode();
                current.getProperties().put(part, newNode);
                current = newNode;
            } else {
                current = (DataNode.ObjectNode) child;
            }
        }

        // Set the value on the final node
        String lastPart = parts[parts.length - 1];

        // Convert value to DataNode if needed
        if (value instanceof DataNode) {
            current.getProperties().put(lastPart, (DataNode) value);
        } else {
            current.getProperties().put(lastPart, convertToDataNode(value));
        }
    }

    private DataNode convertToDataNode(Object value) {
        if (value == null) {
            return DataNode.NullValue.INSTANCE;
        } else if (value instanceof String) {
            return new DataNode.StringValue((String) value);
        } else if (value instanceof Number) {
            return new DataNode.NumberValue((Number) value);
        } else if (value instanceof Boolean) {
            return new DataNode.BooleanValue((Boolean) value);
        } else if (value instanceof DataNode) {
            return (DataNode) value;
        } else if (value instanceof Map) {
            DataNode.ObjectNode objNode = new DataNode.ObjectNode();
            for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
                objNode.getProperties().put(
                        entry.getKey().toString(),
                        convertToDataNode(entry.getValue())
                );
            }
            return objNode;
        } else if (value instanceof Collection) {
            DataNode.ArrayNode arrayNode = new DataNode.ArrayNode();
            for (Object item : (Collection<?>) value) {
                arrayNode.getElements().add(convertToDataNode(item));
            }
            return arrayNode;
        }

        // For other types, convert to string
        return new DataNode.StringValue(value.toString());
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/structure/NestingMappingRule.kt
package xyz.mahmoudahmed.dsl.structure

import xyz.mahmoudahmed.adapter.DataNode
import xyz.mahmoudahmed.dsl.core.MappingContext
import xyz.mahmoudahmed.dsl.core.MappingExecutionException
import xyz.mahmoudahmed.dsl.core.MappingRule
import xyz.mahmoudahmed.dsl.core.SimpleMapping

/**
 * Rule that nests flat fields into a structured object or array.
 * Supports pattern matching for grouping related fields.
 */
class NestingMappingRule(
    private val sourcePattern: String,
    private val targetPath: String,
    private val structureName: String,
    private val fieldNameExtractor: String?,
    private val valueTransformation: ((String, Any) -> Any)?,
    private val isCollection: Boolean,
    private val includeNullValues: Boolean = true, // Changed default to true to include null values
    private val strictTypeHandling: Boolean
) : MappingRule {

    override fun apply(context: MappingContext, target: Any) {
        if (target !is DataNode.ObjectNode) {
            throw MappingExecutionException("Target must be a DataNode.ObjectNode")
        }

        try {
            val sourceNode = context.sourceData as? DataNode
                ?: throw MappingExecutionException("Source data must be a DataNode")

            // Extract pattern type and prepare for matching
            val patternType = analyzePatternType(sourcePattern)

            // Find all paths that match the pattern
            val matchedPaths = findAllMatchingPaths(sourceNode, sourcePattern, patternType)

            if (matchedPaths.isEmpty()) {
                return // Nothing to nest
            }

            // Create the appropriate nested structure
            if (isCollection) {
                createNestedCollection(matchedPaths, target, context)
            } else {
                createNestedObject(matchedPaths, target, context)
            }
        } catch (e: Exception) {
            if (e is MappingExecutionException) throw e
            throw MappingExecutionException(
                "Error creating nested structure for pattern '$sourcePattern': ${e.message}",
                e
            )
        }
    }

    /**
     * Represents the type of pattern we're dealing with.
     */
    private enum class PatternType {
        PREFIX_BASED,
        ID_FIELD_BASED,
        PATH_WILDCARD,
        COMPLEX_WILDCARD
    }

    /**
     * Determines the type of pattern we're dealing with.
     */
    private fun analyzePatternType(pattern: String): PatternType {
        return when {
            // ID_FIELD_BASED: patterns like "prefix_*_*" or "item_*_field"
            pattern.contains("_*_") -> PatternType.ID_FIELD_BASED

            // PREFIX_BASED: patterns like "prefix_*"
            pattern.contains("_*") && !pattern.contains("_*_") -> PatternType.PREFIX_BASED

            // PATH_WILDCARD: patterns like "path.*.field"
            pattern.contains(".*") && !pattern.contains("_*") -> PatternType.PATH_WILDCARD

            // COMPLEX_WILDCARD: any other pattern with wildcards
            pattern.contains("*") -> PatternType.COMPLEX_WILDCARD

            // Default to PREFIX_BASED for patterns without wildcards
            else -> PatternType.PREFIX_BASED
        }
    }

    /**
     * Represents a matched path and its components.
     */
    private data class MatchedPathInfo(
        val path: String,           // Full path
        val value: DataNode,        // Value at the path
        val identifier: String,     // Identifier for grouping (e.g., "1" from "item_1_name")
        val fieldName: String       // Field name (e.g., "name" from "item_1_name")
    )

    /**
     * Finds all paths in the source data that match the given pattern.
     */
    private fun findAllMatchingPaths(
        rootNode: DataNode,
        pattern: String,
        patternType: PatternType
    ): List<MatchedPathInfo> {
        val result = mutableListOf<MatchedPathInfo>()

        // Create a regex for path matching
        val patternRegex = createMatchingRegex(pattern)

        // Recursively traverse the data structure
        traverseNode(rootNode, "", patternRegex, patternType, result)

        return result
    }

    /**
     * Creates a regex for matching paths against the pattern.
     */
    private fun createMatchingRegex(pattern: String): Regex {
        // Handle pattern segments
        val segments = pattern.split(".")
        val regexParts = segments.map { segment ->
            if (segment == "*") {
                // Match any path segment
                "[^.]+"
            } else if (segment.contains("*")) {
                // Replace * with a wildcard in the segment
                segment.replace("*", ".*?")
            } else {
                // Exact match for this segment
                segment
            }
        }

        // Join with escaped dots
        val regexPattern = regexParts.joinToString("\\.").replace("[", "\$$").replace("]", "\$$")

        return Regex("^$regexPattern$")
    }

    /**
     * Recursively traverses the node structure to find matching paths.
     */
    private fun traverseNode(
        node: DataNode,
        currentPath: String,
        patternRegex: Regex,
        patternType: PatternType,
        result: MutableList<MatchedPathInfo>
    ) {
        // Check if current path matches the pattern
        if (patternRegex.matches(currentPath) && currentPath.isNotEmpty()) {
            // Extract components based on pattern type
            val (identifier, fieldName) = extractComponents(currentPath, patternType)
            result.add(MatchedPathInfo(currentPath, node, identifier, fieldName))
        }

        // Continue traversal for complex nodes
        when (node) {
            is DataNode.ObjectNode -> {
                for ((key, value) in node.properties) {
                    val newPath = if (currentPath.isEmpty()) key else "$currentPath.$key"
                    traverseNode(value, newPath, patternRegex, patternType, result)
                }
            }
            is DataNode.ArrayNode -> {
                for (i in node.elements.indices) {
                    val newPath = "$currentPath[$i]"
                    traverseNode(node.elements[i], newPath, patternRegex, patternType, result)
                }
            }
            else -> {
                // Leaf node, no further traversal needed
            }
        }
    }

    /**
     * Extracts identifier and field name from a path based on pattern type.
     */
    private fun extractComponents(path: String, patternType: PatternType): Pair<String, String> {
        val lastSegment = path.substringAfterLast('.')

        return when (patternType) {
            PatternType.PREFIX_BASED -> {
                // For patterns like "personal_firstName", extract "firstName"
                if (lastSegment.contains("_")) {
                    val prefix = lastSegment.substringBefore('_')
                    val fieldName = lastSegment.substringAfter('_')
                    Pair("default", fieldName)
                } else {
                    Pair("default", lastSegment)
                }
            }

            PatternType.ID_FIELD_BASED -> {
                // For patterns like "item_1_name", extract "1" and "name"
                val regex = "(\\w+)_(\\d+)_(\\w+)".toRegex()
                val match = regex.find(lastSegment)

                if (match != null && match.groupValues.size > 3) {
                    val id = match.groupValues[2]
                    val field = match.groupValues[3]
                    Pair(id, field)
                } else {
                    // Try alternate format: "prefix_id_field"
                    val parts = lastSegment.split("_")
                    if (parts.size >= 3) {
                        Pair(parts[1], parts[2])
                    } else {
                        Pair("default", lastSegment)
                    }
                }
            }

            PatternType.PATH_WILDCARD -> {
                // For patterns like "data.*.value", use the wildcard segment as identifier
                val segments = path.split(".")
                val wildcardIndex = sourcePattern.split(".").indexOfFirst { it == "*" }

                if (wildcardIndex >= 0 && wildcardIndex < segments.size) {
                    Pair(segments[wildcardIndex], lastSegment)
                } else {
                    Pair("default", lastSegment)
                }
            }

            PatternType.COMPLEX_WILDCARD -> {
                // For more complex patterns, try to extract meaningful parts
                if (lastSegment.contains("_")) {
                    val parts = lastSegment.split("_")
                    if (parts.size >= 3 && parts[1].matches("\\d+".toRegex())) {
                        // Pattern like "prefix_123_field"
                        Pair(parts[1], parts[2])
                    } else if (parts.size >= 2) {
                        // Pattern like "prefix_field"
                        Pair("default", parts[1])
                    } else {
                        Pair("default", lastSegment)
                    }
                } else {
                    Pair("default", lastSegment)
                }
            }
        }
    }

    /**
     * Creates a nested collection structure from the matched paths.
     */
    private fun createNestedCollection(
        matchedPaths: List<MatchedPathInfo>,
        targetNode: DataNode.ObjectNode,
        context: MappingContext
    ) {
        // Group paths by their identifier
        val groupedPaths = matchedPaths.groupBy { it.identifier }

        // Create collection node
        val collectionNode = DataNode.ArrayNode()

        // Sort groups by identifier (numerically if possible)
        val sortedGroups = groupedPaths.toSortedMap(compareBy {
            it.toIntOrNull() ?: Int.MAX_VALUE
        })

        // Process each group
        for ((_, paths) in sortedGroups) {
            val itemNode = DataNode.ObjectNode()

            // Add each field to the item
            for (pathInfo in paths) {
                val propertyName = pathInfo.fieldName
                val value = pathInfo.value

                // Transform the value if needed
                val transformedValue = transformValue(propertyName, value, context)

                // Add to the item, including null values if configured
                if (transformedValue != null || includeNullValues) {
                    itemNode.properties[propertyName] = transformedValue ?: DataNode.NullValue
                }
            }

            // Only add non-empty items
            if (itemNode.properties.isNotEmpty()) {
                collectionNode.elements.add(itemNode)
            }
        }

        // Set the collection in the target
        val targetPathWithName = if (targetPath.isEmpty()) structureName else "$targetPath.$structureName"
        setValueInPath(targetNode, targetPathWithName, collectionNode)
    }

    /**
     * Creates a nested object structure from the matched paths.
     */
    private fun createNestedObject(
        matchedPaths: List<MatchedPathInfo>,
        targetNode: DataNode.ObjectNode,
        context: MappingContext
    ) {
        // Create object node
        val objectNode = DataNode.ObjectNode()

        // Add each field to the object
        for (pathInfo in matchedPaths) {
            val propertyName = pathInfo.fieldName
            val value = pathInfo.value

            // Transform the value if needed
            val transformedValue = transformValue(propertyName, value, context)

            // Add to the object, including null values if configured
            if (transformedValue != null || includeNullValues) {
                objectNode.properties[propertyName] = transformedValue ?: DataNode.NullValue
            }
        }

        // Only set non-empty objects
        if (objectNode.properties.isNotEmpty()) {
            val targetPathWithName = if (targetPath.isEmpty()) structureName else "$targetPath.$structureName"
            setValueInPath(targetNode, targetPathWithName, objectNode)
        }
    }

    /**
     * Transforms a value using the provided transformation function.
     */
    private fun transformValue(propertyName: String, value: DataNode, context: MappingContext): DataNode? {
        if (valueTransformation == null) {
            return value
        }

        try {
            val transformedValue = valueTransformation.invoke(propertyName, value)
            return when (transformedValue) {
                is DataNode -> transformedValue
                else -> SimpleMapping.convertToDataNode(transformedValue)
            }
        } catch (e: Exception) {
            if (strictTypeHandling) {
                throw MappingExecutionException(
                    "Error transforming value for property '$propertyName': ${e.message}",
                    e
                )
            }
            return value
        }
    }

    /**
     * Sets a value at the specified path in the target node.
     */
    private fun setValueInPath(targetNode: DataNode.ObjectNode, path: String, value: DataNode) {
        try {
            SimpleMapping.setValueInDataNode(targetNode, path, value)
        } catch (e: Exception) {
            throw MappingExecutionException(
                "Error setting nested value at path '$path': ${e.message}",
                e
            )
        }
    }
}